File: src/main/java/core/procedures/strings/Split.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;

import core.scm.Vector;
import java.util.regex.Pattern;

public final class Split extends AFn {

  public Split() {
    super(new FnArgsBuilder().min(2).max(3).mandatory(new Class[]{CharSequence.class, Pattern.class}).rest(Long.class).build());
  }

  @Override
  public String getName() {
    return "split";
  }

  @Override
  public Vector apply(Object... args) {
    if (args.length == 2) {
      return new Vector(((Pattern)args[1]).split((CharSequence) args[0]));
    }
    return new Vector(((Pattern)args[1]).split((CharSequence) args[0], ((Long)args[2]).intValue()));
  }
}

File: src/main/java/core/procedures/strings/Split.kt
package core.procedures.strings

import core.procedures.AFn
import core.procedures.Arity.Range

import core.scm.Vector
import java.util.regex.Pattern

class Split : AFn<Any?, Vector>(name = "split", isPure = true, arity = Range(2, 3),
                                mandatoryArgsTypes = arrayOf(CharSequence::class.java, Pattern::class.java),
                                restArgsType = Long::class.java) {

    override operator fun invoke(args: Array<out Any?>) = when (args.size) {
        2    -> Vector((args[1] as Pattern).split(args[0] as CharSequence) as Array<Any?>)
        else -> Vector((args[1] as Pattern).split(args[0] as CharSequence, (args[2] as Long).toInt()) as Array<Any?>)
    }
}
